### ISSUE TITLE
Fixed seeds do not work

### ISSUE THREAD

**Issue by Maggione** [OpenNMT/OpenNMT-py#392]
I set the "opt.seed" as a positive int value, however, the output results are different each time without changing anything. How can I reproduce the result of models?

**Comment by JianyuZhan** [comment]
Are you training using GPU?  If yes, you probably need to use this `torch.cuda.manual_seed`,  see http://pytorch.org/docs/master/cuda.html?highlight=torch%20manual_seed#torch.cuda.manual_seed

If that works, I will change the code. Thanks.

**Comment by Maggione** [comment]
Yes, it has been used in train.py as follow.
if opt.gpuid:
    cuda.set_device(opt.gpuid[0])
    if opt.seed > 0:
        torch.cuda.manual_seed(opt.seed)
It does not work.

**Comment by srush** [comment]
Huh.. Is this a pytorch issue? What else should we be doing to make sure the seed is used?

**Comment by Maggione** [comment]
I think it maybe the seed of torchtext.data.Iterator is not fixed? Do you know how to fix it?

**Comment by Maggione** [comment]
Yes, it is the problem of IO.py. Because torchtext.data.Iterator shuffle the data by the package "random". You can insert the code:
`import random
random.seed([your seed])`
It works.

**Comment by srush** [comment]
Oh that's a good catch. Can you send a PR?

**Comment by JianyuZhan** [comment]
#436

**Comment by xinyadu** [comment]
randomness still exist, after setting the seed for ``random'' in train.py

**Comment by fuzihaofzh** [comment]
Seems still not work due to the torchtext problem.

**Comment by DianeBouchacourt** [comment]
Same, it does not seem to work for me either. Setting the random.seed() does not change a seed. It would be great to be able to control the order of the samples drawn by the iterator at each epoch for reproducibility.
